pkg/cluster/internal/create/nodes.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package create
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"sigs.k8s.io/kind/pkg/cluster/config"
+	"sigs.k8s.io/kind/pkg/cluster/constants"
+	"sigs.k8s.io/kind/pkg/cluster/nodes"
+	logutil "sigs.k8s.io/kind/pkg/log"
+)
+
+// provisioning order for nodes by role
+var defaultRoleOrder = []string{
+	constants.ExternalLoadBalancerNodeRoleValue,
+	constants.ExternalEtcdNodeRoleValue,
+	constants.ControlPlaneNodeRoleValue,
+	constants.WorkerNodeRoleValue,
+}
+
+// sorts nodes for provisioning
+func sortNodes(nodes []config.Node, roleOrder []string) {
+	roleToOrder := makeRoleToOrder(roleOrder)
+	sort.SliceStable(nodes, func(i, j int) bool {
+		return roleToOrder(string(nodes[i].Role)) < roleToOrder(string(nodes[j].Role))
+	})
+}
+
+// helper to convert an ordered slice of roles to a mapping of provisioning
+// role to provisioning order
+func makeRoleToOrder(roleOrder []string) func(string) int {
+	orderMap := make(map[string]int)
+	for i, role := range roleOrder {
+		orderMap[role] = i
+	}
+	return func(role string) int {
+		p, ok := orderMap[role]
+		if !ok {
+			return 10000
+		}
+		return p
+	}
+}
+
+// TODO(bentheelder): eliminate this when we have v1alpha3
+func convertReplicas(nodes []config.Node) []config.Node {
+	out := []config.Node{}
+	for _, node := range nodes {
+		replicas := int32(1)
+		if node.Replicas != nil {
+			replicas = *node.Replicas
+		}
+		for i := int32(0); i < replicas; i++ {
+			outNode := node.DeepCopy()
+			outNode.Replicas = nil
+			out = append(out, *outNode)
+		}
+	}
+	return out
+}
+
+// provisionNodes takes care of creating all the containers
+// that will host `kind` nodes
+func provisionNodes(
+	status *logutil.Status, cfg *config.Config, clusterName, clusterLabel string,
+) (nodesByName map[string]*nodes.Node, err error) {
+	nodesByName = map[string]*nodes.Node{}
+
+	// convert replicas to normal nodes
+	// TODO(bentheelder): eliminate this when we have v1alpha3
+	configNodes := convertReplicas(cfg.Nodes)
+	// TODO(bentheelder): allow overriding defaultRoleOrder
+	sortNodes(configNodes, defaultRoleOrder)
+
+	nameNode := makeNodeNamer(clusterName)
+
+	// provision all nodes in the config
+	// TODO(bentheelder): handle implicit nodes as well
+	for _, configNode := range configNodes {
+		// name the node
+		name := nameNode(string(configNode.Role))
```

Comment:
The counter is an internal detail to producing the names and could change, the caller doesn't care about the counter, just that it will vend node names. 🙃